### PR TITLE
resolve issue where noise calculation was failing due to incorrect input file

### PR DESCRIPTION
**Pull request by efidoman**

### CODE REVIEW

**Comment by scottleibrand** [comment]
Per discussion at https://gitter.im/thebookins/xdrip-js?at=5d121059260d3b49815dd1bc, we should comprehensively categorize all statuses as heavy noise (don't loop), no noise (loop normally), and light noise for everything else (loop with a higher target).

**Comment by efidoman** [comment]
I found a root issue with the native noise calculation. I also added  the tiered noise failsafe for filtered/unfiltered variation. Let's get these in first and then consider the tx/sensor status tiered noise changes.

**Comment by efidoman** [comment]
Also, g5-noise.sh was an extra file that is not used so I removed it to avoid any confusion

File: xdrip-get-entries.sh [hunk 1]
```diff
@@ -1390,16 +1379,18 @@ function calculate_noise()
...
       noiseSend=4  
       noiseString="Heavy"
-      log "setting noise to heavy because tx is in temporary session error"
+      log "setting noise to $noiseString because - filtered/unfiltered variation of $variation exceeds 35%"
+  elif [ $(bc <<< "$variation >= 30") -eq 1 -o  $(bc  <<< "$variation <= -30") -eq 1 ]; then
+      noiseSend=3  
+      noiseString="Medium"
+      log "setting noise to $noiseString because - filtered/unfiltered variation of $variation exceeds 30%"
+  elif [ $(bc <<< "$variation >= 25") -eq 1 -o  $(bc  <<< "$variation <= -25") -eq 1 ]; then
+      noiseSend=3
```
Comment (scottleibrand) [inline xdrip-get-entries.sh]: Should this be 1?

Comment (efidoman) [reply xdrip-get-entries.sh]: Yes, good catch

Comment (efidoman) [reply xdrip-get-entries.sh]: actually, it should be 2 for Light noise

**Comment by scottleibrand** [comment]
I like the 35%/30%/25% tiers.  I think we should also ensure that anything other than a "normal" or a normal every-12h "calibration required" status gets reported as light noise, so targets can be appropriately raised.

**Comment by efidoman** [comment]
OK - in this PR (or a follow-up one)?

**Comment by tynbendad** [comment]
Did you want to fix variation denominator to a min of the two values?

**Comment by scottleibrand** [comment]
I'm fine with either this or a new PR, depending on which would be easier to get going ASAP.

**Comment by efidoman** [comment]
OK - I will merge this PR after one more test suggested by @tynbendad and then open a new one for the added noise logic based on status. That will be fastest because the status-based noise will take some extra work and testing to ensure it's done right.

**Comment by efidoman** [comment]
Added back g5-noise.sh since it is a utility used to report the latest noise from the command line.
Also, I'm doing a global-install to make sure none of these changes cause an install issue.

**Comment by efidoman** [comment]
1st test of @scottleibrand data gives me no noise by native calculations, but will give noise by variation method. I think another thing to consider for noise is that we'll have 3 methods for getting noise and we need to take the highest noise of the 3 noise checks (native, filtered/unfiltered variation, and status). That should also be in the follow-up PR.

root@eddie1:~# cat n*test1.csv
1561629533,77,77,87
1561629833,72,72,75
1561630132,74,72,81
1561630432,107,77,121
root@eddie1:~# cgm-calc-noise n*test1.csv
[{"noise":0.00000}]

**Comment by efidoman** [comment]
Testing native noise - there is a time-based factor so let me evaluate if that affects my testing.

**Comment by efidoman** [comment]
2nd test also no noise ... hmm.
```
root@eddie1:~# cat noise2.csv
1561629533,77,77,87
1561629833,72,72,75
1561630132,74,72,81
1561630432,107,77,121
1561630733,81,84,101
root@eddie1:~# cgm-calc-noise noise2.csv
[{"noise":0.00000}]
```

**Comment by efidoman** [comment]
In the 3rd test I modified @scottleibrand data to simulate more noise and got a small value for noise. The logic needs adjusting. It used to calculate noise better so I'm going to look at the history of changes for this g5-noise-calc.sh

```
root@eddie1:~# cp noise2.csv noise3.csv
root@eddie1:~# vi *3.csv
root@eddie1:~# cat noise3.csv
1561629533,77,77,87
1561629833,72,72,75
1561630132,74,72,81
1561630432,167,177,151
1561630733,81,84,101
root@eddie1:~# cgm-calc-noise noise3.csv
[{"noise":0.00005}]
```

**Comment by efidoman** [comment]
I found the change that broke the noise input file logic, but not any change that modified the algorithm. Doing more tests now.

**Comment by efidoman** [comment]
It is still not giving noise value to amount to much of anything. This worked before. My test must have some issue. Going to uncomment some logging.

```
root@eddie1:~# cat noise4.csv
1561629533,77,77,87
1561629833,72,72,75
1561630132,74,72,81
1561630432,167,177,151
1561630733,81,84,81
1561630933,121,132,141
root@eddie1:~# cgm-calc-noise noise4.csv
[{"noise":0.00004}]
```

**Comment by efidoman** [comment]
That didn't help me much.

```
root@eddie1:~# cgm-calc-noise noise4.csv
0 9000 17970 26970 36000 42000
1561629533 1561629833 1561630132 1561630432 1561630733 1561630933
77 72 74 167 81 121
yDelta=-5.20833333333333333330, xDelta=9000
yDelta=2.81666666666666666665, xDelta=8970
yDelta=104.62500000000000000000, xDelta=9000
yDelta=-130.43333333333333333258, xDelta=9030
yDelta=62.83333333333333333316, xDelta=6000
sod=42001.88102422164960987730, overallDistance=42000.02304761272389937772
noise = .00004423555716121982
[{"noise":0.00004}]
root@eddie1:~# 
```

**Comment by efidoman** [comment]
Yes, the calculation is time-based. I need to use valid time values in my tests. overallDistance is in the numerator. I will need to use my valid input file and modify the numbers to simulate noise. I do remember that's how I tested it before.

**Comment by efidoman** [comment]
Much better ...
 
My tests were bad, now to redo them

```
root@eddie1:~# cgm-calc-noise noise-input41.csv
0 -72180 -63180 -54180 -45180 -36180 -27180 -18150 -9180 0
1561635538 1561633132 1561633432 1561633732 1561634032 1561634332 1561634632 1561634933 1561635232 1561635538
73 null 72 71 71 70 170 71 172 73
yDelta=-74.82500000000000000000, xDelta=-72180
yDelta=98.28000000000000000000, xDelta=9000
yDelta=-1.39750000000000000000, xDelta=9000
yDelta=0, xDelta=9000
yDelta=-1.12500000000000000000, xDelta=9000
yDelta=149.50000000000000000000, xDelta=9000
yDelta=-151.22250000000000000000, xDelta=9030
yDelta=157.56000000000000000000, xDelta=8970
yDelta=-157.65750000000000000000, xDelta=9180
sod=144365.82068624511894400425, overallDistance=0
noise = 1.00000000000000000000
[{"noise":1.00000}]

```

**Comment by efidoman** [comment]
@scottleibrand - can you help look at the logic in g5-calc-noise.sh? It is documented well and used to work, but I can't figure out why overallDistance is zero here.

**Comment by efidoman** [comment]
I do think that the multiply by 30 to normalize the time axis is wrong now since the filtered/unfiltered was divided by 1000 a while back for readability / consistency.

**Comment by efidoman** [comment]
Getting better now.

```
root@eddie1:~# cat *41*
1561634332,70,71,93
1561634632,70,71,93
1561634933,71,70,94
1561635232,72,71,95
1561635533,73,71,97
1561635832,75,72,99
1561636132,176,174,170
1561636432,77,75,101
1561636739,178,177,172
root@eddie1:~# cgm-calc-noise noise-input41.csv
[{"noise":0.03847}]
root@eddie1:~# 
```

**Comment by efidoman** [comment]
Better after I calibrated the x-axis (time to be in the same ballpark as the y-axis (BG values).

```
root@eddie1:~# cgm-calc-noise noise-input41.csv
0 100.00000000000000000000 200.33333333333333333333 300.00000000000000000000 400.33333333333333333333 500.00000000000000000000 600.00000000000000000000 700.00000000000000000000 802.33333333333333333333
1561634332 1561634632 1561634933 1561635232 1561635533 1561635832 1561636132 1561636432 1561636739
70 70 71 72 73 75 176 77 178
[{"noise":0.21272}]
root@eddie1:~#

**Comment by efidoman** [comment]
Good questions. The basic premise of noise in native calculation is to compare the start point to the end point calculating a straight line, thus the sum of the sqrts. Then noise is a ratio of that shortest distance to the overall distance traveled by bg over that same x-axis time period. This ratio will get lower as bg is stable and higher as bg is bouncy. Thus the reason for normalizing and subtracting from 1 to get a noise percentage that is lowest noise at 0.00 and highest noise at 1.00

**Comment by efidoman** [comment]
It's working now ... the 2nd issue was the time axis wasn't modified to incorporate the fact that filtered and unfiltered were divided by 1000 several revisions back. Now, I can just modify the x-axis factor to amplify the noise signal to where we want it. the shorter the x axis, the more bouncy bg shows noise.

**Comment by efidoman** [comment]
So right now (with it working for me) there are 300 seconds between each bg value and I'm dividing by 3 to get basically 100 for x axis. If bg is average 100, then that means the x and y axis are squarely normalized.

This is what I strived to accomplish originally when the thresholds were set, so let's test out various noise scenarios with these changes including @scottleibrand noise scenario and see how it calculates. We may not need to adjust the x-factor again. I'll try to document in the code how the x-factor works more once we fine-tune the x-factor.

**Comment by efidoman** [comment]
Capture

**Comment by efidoman** [comment]
It's working ... look at that - it gave light noise when my bg rose very quickly by 22 points. My sensor is nearing end of life.

**Comment by efidoman** [comment]
Next bg is clean as expected.
Capture

I'll do more detailed testing tonight.

**Comment by scottleibrand** [comment]
Sounds like your noise algorithm is more complicated than necessary if it includes the x axis and then removes it again. If you’re looking for a “length of line” algorithm, the contribution of the x axis will solely be a function of the time interval you’re trying to sample. As a result, if you have readings for the entire time interval, you can instead simply look at the square of the sum of the BG deltas.

**Comment by efidoman** [comment]
The reason the time axis is included was because we needed a way to give more weight to the most recent BG noise. The noise is amplified for the most recent deltas. Maybe that can still be accomplished and simplify at the same time by removing the x-axis as you suggest. That level of refactoring may be more than we want to take on with this PR?

**Comment by efidoman** [comment]
If we don't amplify the most recent noise, then the following series will give the same noise level which is not what I would expect ...

100, 200, 100, 101, 102, 103, 104, 105, 106, 107
101, 102, 103, 104, 105, 106, 107, 100, 200, 100

I would expect the first series to give light or no noise and the second line of BG values to give heavy noise.

Maybe there's a much more simple way to handle this. I'm open to suggestions or anyone offering a better more simple way.

**Comment by tynbendad** [comment]
I like the noise coming down quicker so calc with weighting sounds good to me, I just didn't understand it.

**Comment by scottleibrand** [comment]
I agree that refactoring the noise calculation should be a separate PR.

**Comment by efidoman** [comment]
Sounds good. I'll finish testing this one, merge it, and then immediately open a noise refactor PR.

**Comment by efidoman** [comment]
It isn't yet where it needs to be. The filtered/unfiltered variation noise scale is also not doing what we want it to do. See below where it is putting light and medium noise when I'm having a normal rise from a meal.

```
1561655636,2019-06-27 13:13,66,54,FortyFiveUp,89,79,,0.812,-6.611,0,4,0.98196,0.01,1,expired,1.00,1,-78
1561655936,2019-06-27 13:18,75,56,DoubleUp,100,85,,0.812,-6.611,0,4,0.98196,0.01,3,expired,1.00,1,-74
1561656236,2019-06-27 13:23,83,64,SingleUp,110,101,,0.812,-6.611,0,4,0.98196,0.01,2,expired,1.00,1,-69
1561656536,2019-06-27 13:28,97,77,SingleUp,127,127,,0.812,-6.611,0,4,0.98196,0.01,2,expired,1.00,1,-83
1561656836,2019-06-27 13:33,111,91,DoubleUp,144,163,,0.812,-6.611,0,4,0.98196,0.01,1,expired,1.00,1,-81
```

**Comment by efidoman** [comment]
The native noise calc is not good enough either yet.

Here's using some of Scott's data:

```
root@eddie1:~# cgm-calc-noise ./test6.csv
0 100.00000000000000000000 199.66666666666666666666 300.00000000000000000000 400.00000000000000000000 500.00000000000000000000 600.00000000000000000000 700.00000000000000000000 801.33333333333333333333
1561721330 1561721630 1561721929 1561722230 1561722530 1561722830 1561723130 1561723430 1561723734
72 69 65 83 77 72 74 107 81
[{"noise":0.02364}]
root@eddie1:~# cat test6.csv
1561721330,72,74,96
1561721630,69,74,93
1561721929,65,71,88
1561722230,83,70,110
1561722530,77,77,87
1561722830,72,72,81
1561723130,74,72,810
1561723430,107,77,121
1561723734,81,84,101
root@eddie1:~#
```

**Comment by efidoman** [comment]
I will tweak the x-axis once more and try again

**Comment by efidoman** [comment]
Alright - using a divide by 20 for the x-axis gets it to show noise with this data set. I'll run with this today while the sensor is still supposed to be dying soon. 

```
root@eddie1:~/src/Logger/bin# cd
root@eddie1:~# cgm-calc-noise ./test6.csv
0 15.00000000000000000000 29.95000000000000000000 45.00000000000000000000 60.00000000000000000000 75.00000000000000000000 90.00000000000000000000 105.00000000000000000000 120.20000000000000000000
1561721330 1561721630 1561721929 1561722230 1561722530 1561722830 1561723130 1561723430 1561723734
72 69 65 83 77 72 74 107 81
[{"noise":0.38248}]
```

**Comment by efidoman** [comment]
Reviewing the last several hours of my dying sensor and noise calcs are much better. I will merge this in shortly and then open the refactor PR.

```
Date,          unfilt, filt, LSRbg, Txbg, noise, noisesend, noisetext
20190628-14:53, 178,    192,     227,    0,    0.64,    3,    Medium
20190628-14:58, 188,    195,     239,    0,    0.63,    3,    Medium
20190628-15:03, 196,    188,     249,    0,    0.56,    3,    Medium
20190628-15:08, 136,    179,     175,    0,    0.41,    4,    Heavy
20190628-15:13, 98,    163,     128,    0,    0.69,    2,    Light
20190628-15:18, 93,    135,     122,    0,    0.69,    1,    Clean
20190628-15:23, 206,    116,     261,    0,    0.68,    4,    Heavy
20190628-15:28, 201,    129,     255,    0,    0.66,    4,    Heavy
20190628-15:33, 194,    168,     247,    0,    0.59,    3,    Medium
20190628-15:38, 189,    203,     240,    0,    0.50,    3,    Medium
20190628-15:43, 185,    210,     235,    0,    0.44,    3,    Medium
20190628-15:48, 173,    196,     221,    0,    0.06,    3,    Medium
20190628-15:53, 166,    178,     212,    0,    0.06,    2,    Light
20190628-15:58, 153,    166,     196,    0,    0.06,    1,    Clean
```

**Comment by efidoman** [comment]
BTW - I updated the cgm-noise command line utility to do much better noise reporting from the cgm.csv logfile.

**Pull request merged**